push-notifications.js

```diff
@@ -89,64 +139,156 @@ PushNotifications.prototype.publish = function(interests, publishRequest) {
+            );
+        }
+    }
+
+    publishRequest.users = users;
+    const body = JSON.stringify(publishRequest);
+    const options = {
+        uri: `${this.protocol}://${this.endpoint}/publish_api/v1/instances/${
+            this.instanceId
+        }/publishes/users`,
+        method: 'POST',
+        body,
+        headers: {
+            'Content-Type': 'application/json',
+            'Content-Length': Buffer.byteLength(body)
         }
     };
 
-    return doRequest(payload, options);
+    return doRequest(options);
 };
 
-function doRequest(payload, options) {
-    const promise = new Promise((resolve, reject) => {
-        const request = https.request(options, response => {
-            const wasSuccessful = response.statusCode === 200;
-            let responseString = '';
-            response.setEncoding('utf8');
-
-            response.on('data', function(chunk) {
-                responseString += chunk;
-            });
-            response.on('end', function() {
-                let responseBody;
-                try {
-                    responseBody = JSON.parse(responseString);
-                } catch (_) {
-                    reject(
-                        new Error('Unknown error - invalid server response')
+PushNotifications.prototype.deleteUser = function(userId) {
+    if (userId === undefined || userId === null) {
+        return Promise.reject(new Error('User ID argument is required'));
+    }
+    if (typeof userId !== 'string') {
+        return Promise.reject(new Error('User ID argument must be a string'));
+    }
+    if (userId.length > USERS_STRING_MAX_LENGTH) {
+        return Promise.reject(new Error('User ID argument is too long'));
+    }
+
+    const options = {
+        uri: `${this.protocol}://${this.endpoint}/user_api/v1/instances/${
+            this.instanceId
+        }/users/${encodeURIComponent(userId)}`,
+        method: 'DELETE'
+    };
+
+    return doRequest(options);
+};
+
+function isJsonString(str) {
+    try {
+        JSON.parse(str);
+    } catch (e) {
+        return false;
+    }
+    return true;
+}
+function isValidJson(value) {
+    if (typeof value === 'object' || typeof value === 'number') {
+        return true;
+    }
+    if (typeof value === 'string') {
+        return isJsonString(value);
+    }
+    return false;
+}
+
+function doRequest(options) {
+    return baseRequest(options)
+        .then(res => {
+            if (res.body) {
+                return isValidJson(res.body)
+                    ? Promise.resolve(res.body)
+                    : Promise.reject(
+                          new Error('Could not parse response body')
+                      );
+            }
+            return Promise.resolve();
+        })
+        .catch(err => {
+            if (err.name === 'RequestError' || err.name === 'TransformError') {
+                return Promise.reject(err.error);
+            }
+            if (err.name === 'StatusCodeError') {
+                if (
```

Comment:
What if the error response has an incorrect schema?
We could return a nicer error message here